### PR TITLE
Add float type support

### PR DESCRIPTION
**Pull request by ulugbekna**
- added float type support
- minor refactor in `sqrt` function for readability

### CODE REVIEW

**Comment by ulugbekna** [comment]
This overlaps with the previous pr, but I wanted them to be separate, as they have separate concerns.

File: src/efftester.ml [hunk 1]
```diff
@@ -525,11 +539,11 @@ let typeGen = (* Generates ground types (sans type variables) *)
   Gen.fix
     (fun recgen n ->
        if n = 0
-       then Gen.oneofl [Unit;Int;Bool;String]
+       then Gen.oneofl [Unit;Int;Float;Bool;String]
        else
          Gen.frequency [
            (* Generate no alphas *)
-           (4,Gen.oneofl [Unit;Int;Bool;String]);
+           (4,Gen.oneofl [Unit;Int;Float;Bool;String]);
```
Comment (gasche) [inline src/efftester.ml]: We should have a `base_types` variable instead of repeating the same list in several places.

File: src/efftester.ml [hunk 2]
```diff
@@ -955,6 +989,7 @@ let createLit t =
   match t with
   | Unit   -> toTerm LitUnit
   | Int    -> toTerm (LitInt (Gen.generate1 small_int.gen))
+  | Float  -> toTerm (LitFloat (Gen.generate1 float.gen))
```
Comment (gasche) [inline src/efftester.ml]: Note: the `float` generator of QCheck is rather weak (see [source code](https://github.com/c-cube/qcheck/blob/b14291a/src/core/QCheck.ml#L118-L120)), generating a somewhat uniform random float. For a compiler tester you want to generate corner cases more often, such that `0., 1., -1.` to create optimizations opportunities, very small numbers (`Float.epsilon`) to test precision issues, negative zero to see some correctness bugs, `{min,max}_float`, `nan`, and positive/negative infinities.

We should experiment with a better generator locally, but if we have something nice we could also send it to the QCheck upstream.

Comment (gasche) [reply src/efftester.ml]: To clarify: I think that the present generator is fine for the current PR, we can experiment with stronger generators as followup work.

**Review by gasche — COMMENTED** [review]
There seems to be some mismatch between the formatting logic in this PR (using ocp-indent, if I understand correctly) and the other (using ocamlformat), so I guess this will have to be rebased. But overall this looks good and fairly simple, which is nice.

**Comment by jmid** [comment]
I don't use or know ocamlformat, so I'll refrain from commenting on it.
I second @gasche's point about a skewed generator of floats with a greater chance of corner cases.

Have you tried generating programs with
```
  Print.option (toOCaml ~typeannot:false) (Gen.generate1 term_gen.gen);;
```
to observe whether the extension works as intended? 
From what I tell with `string_of_float` and `print_float` there should be some chance of generating float-code in sub-trees. 
The underlying problem here is that `term_gen` is hard-coded to generate an `int` program, which is then let-bound and wrapped in a printer to ensure some output for comparison. As we extend with more and more base types, it would probably be preferable to choose randomly between generating an `int`, `string`, `unit`, `bool`, `float`, ... program, and then wrap the output in a type-directed fashion.

**Comment by gasche** [comment]
For the record I played a bit with the generator, looking at the generated programs, and I noticed the same issue @jmid pointed out that the toplevel only generates an integer value -- see the definitions `term_gen` and `print_wrap`. It would be nice to lift that restriction to better test the floating-point generators.

For my testing I commented out all non-float-related definitions. I also replaced your use of `print_float` with something that does not approximate while printing: `(fun x -> print_string Int64.(to_string (bits_of_float x)))`.

I think the idea of "commenting out all non-flat definitions" corresponds to a common idea in the QuickCheck world, which is that when you have N combinators, it can be more efficient to randomly select a smaller group of K<N combinators and test with only that (for example, "let's only use float operations"). Eventually we should try to get this logic into the program generator, by selecting a (coherent) subset of the initial environment instead of always using the full environment. One easy way to do this would be to define separate environments for separate "features" that we want to test (int, string, float, lists, etc.), and compose the initial environment as a union/concatenation of a randomly-selected subset of feature-specific environments. I may open a separate issue to track this idea for later.

**Comment by gasche** [comment]
At some point it would be nice to rebase this PR on top of the current master (the different reformatting), to get the float operations merged. You ( @ulugbekna ) could decide whether to do that now, or after some of the other things we discussed for the short term.

**Comment by ulugbekna** [comment]
Rebasing is definitely what we would want to do. I will be on it after fixing up some problems noted in the review. 

I was also thinking of closing this PR and creating a new one where I would not have commits involving ocp-indent. This would keep the commit tree cleaner and let us have much less pain in dealing with conflicts.

**Comment by gasche** [comment]
My approach to this problem (where it's not just a rebase, but many commits are not relevant anymore) is to start a clean branch from `master`, get it to feature-parity by cherry-picking commits or even recreating them manually, and then resetting the PR branch to that new branch.

```
# setting up
git checkout master
git checkout -b float-2

# recreating a clean branch
git log add_float_type_support # look at the commits you are interested in
git cherry-pick ... # to reuse a commit from the old branch
<perform some changes>
git commit -C ... # to reuse a commit *message* from the old branch, corresponding to the changes you performed

# swapping the branches
git checkout add_float_type_support
git branch -c add_float_type_support float-old # save the current state in a float-old branch, just in case
git reset --hard float-2  # change add_float_type_support to the state of float-2
# test/check the result
# if something goes wrong, float-old has a copy of the old branch
# if everything went right, you can get rid of float-old and float-2
git branch -D float-old
git branch -d float-2
```

Once `add_float_type_support` contains the new clean branch on top of master, you can `git push -f` to your personal github fork to update the present PR.

**Comment by ulugbekna** [comment]
Now the `effmain` runs `let x = m in n` at the toplevel, where `x` can be of types `int`, `float`, and `string`, `m` is the generated term, and `n` is the printing of variable `x` based on its type. Every run of tester will assign only one type to `x`, i.e., if `x` is initially of type `float`, all tests will run with `x` being a float. This makes test run less comprehensive but can help with modularity. 

I would say that this is a draft implementation, which I really wanted to be out, as I would like to get feedback if I am going in the right direction.

**Comment by gasche** [comment]
That's good with me.

For later: as a generalization of your `printer_by_etype` function, you can in fact define a generic way to observe a value at any type: `observe_ty ty` which generates (randomly) a term of type `ty -> unit`, by matching on `ty`: 
- if `ty` is `unit`, do nothing at all
- if the type is a base type, use the appropriate printing function
- if the type is of the form `ta -> tb`, you can generate a small list of terms of type `ta` and, for each value `x` in that list, apply the `ta -> tb` function to the value and observe the result (at type `tb`).

This is useful at the toplevel of the program, but we could also use it within the generator itself, as a generic construction from any type to `unit`. (This subsumes having many printing functions in the typing environment.)

**Comment by gasche** [comment]
I believe that the PR is in mergeable state now, so I'll merge.

File: src/efftester.ml [hunk 3]
```diff
@@ -1489,4 +1525,19 @@ let eq_test =
       | Some t -> nativeByteEquivalence (toOCaml (print_wrap t)))
 ;;
 
+let rand_eq_test =
+  let typ = Gen.generate1 (Gen.oneofl [ Int; Float; String ]) in
```
Comment (jmid) [inline src/efftester.ml:1529]: I agree that for one top-level invocation, all tests will have the same `typ`.

However I would recommend against this style for reproducability.
In case the generator actually triggers a bug we would want to be able to reproduce the problematic run
with a given seed. You will not immediately be able to so if you mix in`Gen.generate1`, as it kicks off its own random number generator: https://github.com/c-cube/qcheck/blob/master/src/core/QCheck.ml#L255

So what do we do? A simple solution if we want to run separate `int`, `float`, and `string` tests is to write a parameterized test and just invoke it three times.
A more complex solution is to use the `Gen` module's bind operator `>>=` to write a generator of a dependent pair `(typ, term)` such that the type of `term` is `typ`. 

http://janmidtgaard.dk/quickcheck/slides/lec06.pdf  (p.24 - called 20/35)
contains a simple example of such a dependent generator (of numbers).
Here's some code realizing something along the above (with an option type on top to represent failure of the generator's backtracking):
```
let gen =
    Gen.(basetypeGen >>= fun typ ->
         termGen initTriEnv typ >>= fun t_opt -> match t_opt with
         | None -> return None
         | Some t -> return (Some (typ,t)))
```
